altair/utils/_vegafusion_data.py

```diff
@@ -0,0 +1,199 @@
+import sys
+from toolz import curried
+import uuid
+from weakref import WeakValueDictionary
+
+from typing import Union, Dict, Set, MutableMapping
+
+if sys.version_info >= (3, 8):
+    from typing import TypedDict, Final
+else:
+    from typing_extensions import TypedDict, Final
+
+from altair.utils.core import _DataFrameLike
+from altair.utils.data import _DataType, _ToValuesReturnType, MaxRowsError
+from altair.vegalite.data import default_data_transformer
+
+# Temporary storage for dataframes that have been extracted
+# from charts by the vegafusion data transformer. Use a WeakValueDictionary
+# rather than a dict so that the Python interpreter is free to garbage
+# collect the stored DataFrames.
+extracted_inline_tables: MutableMapping[str, _DataFrameLike] = WeakValueDictionary()
```

Comment:
I didn't know about `WeakValueDictionary`, very cool!

In the [docs](https://docs.python.org/3/library/weakref.html) it says:

> Not all objects can be weakly referenced. Objects which support weak references include class instances, functions written in Python (but not in C), instance methods, sets, frozensets, some [file objects](https://docs.python.org/3/glossary.html#term-file-object), [generators](https://docs.python.org/3/glossary.html#term-generator), type objects, sockets, arrays, deques, regular expression pattern objects, and code objects.

Do you know if Pandas/Polars/PyArrow dataframes all support weak references?

Reply:
I believe the "(but not in C)" qualifier is related to native functions, not instances of Python classes that include native implementation.

Here's an example of this working with pandas and polars:

```python
>>> import pandas as pd
>>> import polars as pl
>>> from weakref import WeakValueDictionary
>>> dfs = dict(df1=pd.DataFrame({"a": [1, 2, 3]}), df2=pl.DataFrame({"a": [1, 2, 3]}))
>>> weak_dfs = WeakValueDictionary(df1=dfs["df1"])
>>> list(weak_dfs.keys())
['df1', 'df2']
>>> del dfs["df1"]
>>> list(weak_dfs.keys())
['df2']
>>> del dfs["df2"]
>>> list(weak_dfs.keys())
[]
```

Reply:
Thanks for checking! Looks good to me as well.